### ISSUE TITLE
Centipede trackball weird control

### ISSUE THREAD

**Issue by stiegis** [MiSTer-devel/Arcade-Centipede_MiSTer#23]
Centipede control with trackball is nearly impossible since high score feature was added. The former version without high score did not play well with trackball but with the latest version the trackball does respond even worse.
It takes a few spins to start movement sometimes centipede moves the opposite direction sometimes it feels like it is "floating" there is very little responsiveness and it is hard to control.

The thing I don´t understand is what has the high score feature to do with the behaviour of the trackball? 

Is it possible to improve the behaviour of the trackball? The main problem being responsiveness.

**Comment by JimmyStones** [comment]
> Centipede control with trackball is nearly impossible since high score feature was added. The former version without high score did not play well with trackball but with the latest version the trackball does respond even worse. It takes a few spins to start movement sometimes centipede moves the opposite direction sometimes it feels like it is "floating" there is very little responsiveness and it is hard to control.
> 
> The thing I don´t understand is what has the high score feature to do with the behaviour of the trackball?
> 
> Is it possible to improve the behaviour of the trackball? The main problem being responsiveness.

Hi - could you use the advanced mode in Input Test core to confirm what your trackball appears as in there?  I'm assuming it will be presenting as a mouse but need to confirm so I can work out how to test it!

There shouldn't be anything that directly affected the trackball control in that update, but there were quite a few things changed behind the scenes so it is always possible.

**Comment by stiegis** [comment]
Hello!

I just entered the advanced input test core and can say that it is recognised as mouse. It is stated in the core as mouse and when I spin the trackball the coordinates change according to the movement. As I have already said the trackball is very good to control within the core - the pointer moves as soon as I start spinning an you can position the pointer precisely by stopping the trackball whereas in the centipede core it will slip a littler furter so it is impossible to tell where it will "land" which makes the game kind of unplayable.

In mister when you start a core you can select "keyboard mode" in order to select Joystick 1 or Joystick 2 or mouse etc. I noticed that with centipede you can not select mouse only joystick 1 or 2 perhaps this is the secret to the solution.

Would be super happy if you could fix this

kind regards,

**Comment by JimmyStones** [comment]
Could you give this build a try please?

https://github.com/JimmyStones/Arcade-Centipede_MiSTer/blob/cocktail-fix/releases/Arcade-Centipede_20220405.rbf

I can only test it with the mouse which I didn't find the old version very nice with either, but this seems a little better.  Fingers crossed :)

**Comment by stiegis** [comment]
Hello Jimmy!

Thanks for the test version!

It feels a little better but all in all responsiveness is pretty bad. It does weird things such as accelerate too fast sometimes goes backwards overall you have got the feeling you are not really in control of a prescise movement as you would do with a mouse for example when working in windows. It is kind of "slippery"

**Comment by JimmyStones** [comment]
Ok, disappointing!  Can you give me some details of your trackball please?  Also can you tell me if you have "Fast USB Polling" enable on your MiSTer (and if you have tried with that on or off)

**Comment by stiegis** [comment]
Hello!

I have tested the ball with and without usb polling on and there is no difference. I use a happ trackball - I think it is more or less like the original Atari trackball with Atari/Happ Trackball Optic PCB as encoder .. the leads go into ultimarc optipac and from there into the mister otg usb hub.

Somewhere in this line there must be some funky stuff going on like signals beeing processed twice this would explain the strange behaviour of the trackball movement - on the other hand in input core tester it is perfect like you would expect it from a normal computer mouse (which is a trackball in a way)

**Comment by JimmyStones** [comment]
Ok, I will research further on this - there is another user with Ultimarc hardware that is not experiencing issues, so there must be something funny going on!

**Comment by stiegis** [comment]
Hi Jimmy!

You said that the other guy is using Ultimarc hardware and the trackball is working fine - perhaps he is satisfied with the way it is playing this is a matter of feeling which is hard to compare

However I made a whole new setup of mister last night everything is fresh now and guess what - same behaviour of the trackball .. I am going nuts

Ok you also think something funny is going on - but why does it play ok in input test core when something is wrong in the chain

Can you think there is some wrong setting in the ini file? Does vsync option influence the core in any way. I also tried fast usb polling on and off .. are you sure it is not a problem with the core?

regards,
Alex

**Comment by stiegis** [comment]
Hi!

I made two interesting observations concerning the trackball. Perhaps this helps analyze the problem

Situation 1: I keep rolling the trackball but the character is not moving seems like it is nailed down and takes some time to move again

Situation 2: There is no physical movement of the trackball but some process behind keeps moving the character some distance

Situation 3: The character moves in the opposite direction of the movement of the trackball it takes some time for the character to move again in the right direction

This is what I mean - it is not only that there is little responsiveness the whole thing behaves strangely

What on earth is going on here?

**Comment by JimmyStones** [comment]
It is very unusual!  I have another idea for something to test so I will do a new build after work and send you a link

**Comment by stiegis** [comment]
Hi Jimmy!

Thanks so much for helping me. Highly appreciated.

I am also onto something - I discovered that there is some additional information in the mister ini core which specifys what type of encoder etc mister is connected to therefore you have to give VID a corresponding number for example if it is J-pac or i-pac and there it is also the spinner throttle which I was missing in my original mister ini file. I will set all the additional parameters precisely and see how it behaves then

Also I wanted to disconnect i-pac (joysticks and buttons) and see how the opti-pac behaves only with trackball but this is not simply possible because you need i-pac in order to start the core

My theory is now that there is some input configuration conflict going on that is joystick movements infering with trackball - but I will take care of this as I mentioned above

Perhaps we cannot solve the mystery but we tried and can exclude even more wrong paths

regards

**Comment by JimmyStones** [comment]
Cool, I'll be interested to hear the results!

You can also try this build when you get a chance to see if there is any noticeable difference:

https://github.com/JimmyStones/Arcade-Centipede_MiSTer/blob/cocktail-fix/releases/Arcade-Centipede_20220407.rbf

**Comment by stiegis** [comment]
Hello Jimmy!

This project is a tough one. I messed around with a few setting in the mister ini file and integrated the new core and I have to say that it is definitely better than before .. there is no doubt about it.

Still there is some weird behavoiour going on. The character now reacts much better as soon as you spin the trackingball but when there a few snakes coming down it is not easy to position the character precisely underneath them in order to shoot them. The biggest problem is still the movement in the wrong direction. Say the character rolls left for a while - with a mouse you would immediately stop this by going right but the trackball will still hold on to the left movement for some time. It just feels not even close like when you are moving a mouse pointer in windows. This is what I call precise.

In any case this problem is a complex one that is for sure considering that all the other controls work perfectly fine in mister. On top of that spinner works perfectly in Arkanoid 

Is it possible to ingerate mouse support into the core?

regards

**Comment by JimmyStones** [comment]
Unfortunate fact - this is the mouse support!  Trackballs on MiSTer usually present as a mouse (as you saw in the input test core), and that is how the Centipede core reads it as well.  The difference is that this core has to generate a direction and a pulse (that emulates the way the original Atari trakball worked) from the 'mouse' movement that the CPU reads like on the original hardware.  It's all going back and forth far too many times! 😬

Have you tried it with a mouse?  That's the only way I can test it and it does feel a bit off but not necessarily as bad as it sounds like it is for you.

**Comment by stiegis** [comment]
Hi Jim!

I am going to test it with a mouse. Perhaps this has got to do something with the operating system as well the driver for the mouse can be written for windows or in the case of mister fpga it is linux .. just a thought

I am no technician or programmer by any means but I think I understand the nature of the problem.

For the operating system or the centipede core it is necessary to do a query about the positioning of the trackball that is in physical terms the x and y position. I am aware that there is a whole chain in this process you got the encoder wheel the opti pac mister itself etc. but at the end of the day this is the task it has to fulfill.

Lets assume that the query of the trackball is too slow or too fast (not sure) the it would by nature explain the weird behaviour of the trackball concerning that the calculation in the background is doing its task mathematically speaking correct.

For example when I roll the trackball over and over and the character still stays on place (with a mouse it would no do this) - this is because the calculation of the actual positioning is far behind. After some time it will do so of course but for the game as such this is of course not so good. When this happens and I wait some time and do nothing the "calculation" will catch up and I can move the charcater normal as usual. From this oberservation I come to this conclusion. It is a theory of course but happens in reality.  To put it in a simple way what we can see here is that the real time translation is not happening all the time

So in conclusion the calculation and therefore the timing is not spot on which also explains that the character is moving in the wrong direction sometimes (still calculation the former trackball input). Perhaps the trackball ecoder wheel puts out too many or too little signals (but this is contrary to the motion in mister fpga input core test)

Ok I think I understand it but does this help us?

I will hold on to this subject and continue studying

regards,

**Comment by JimmyStones** [comment]
It's complicated for sure!  I think you are right that it is a timing thing, and maybe the base movement speed the encoder sends is different for different devices so others are not seeing the issue.

I'll have a think and see if I can add a trackball simulator of some kind to the input test core so we can get more diagnostics out to compare.

**Comment by stiegis** [comment]
Hi Jimmy!

I played around with a few settings in the mister ini file without any success. It does not have anything to do with the settings and it is not a hardware issue. I am pretty sure about this

I tried the trackball in the arkanoid core an it works perfectly fine. The thing is it just  feels like when you are using a computer mouse no weird behaviour no direction change or intermittent loss of control and the core has absolute the same exterior paramaters and settings as centipede core would have.

There must be some funky stuff withing the core messing up the timing and positioning of the character. Perhaps you can get some information from the circuit diagram of centipede arcade core.

Is ist possible to compare input test core or akanoid rbf core to centipede in order to see how it is done there or compare to centipede core to see if something is messed up there? From an engineers or programmers point this is way above me. Perhaps you can find out something but in my opinion the cause for this specific issue lies in the core.
The two other cores is proof of working trackball the way it should be (no hardware issue no mister´s fault)

regards,

**Comment by JimmyStones** [comment]
I wrote the Input Test core so I know how that one is done already 😁  Unfortunately it is the most simple case - the mouse movement is just added to a position every frame (60 times per second), no messing around with encoding, so doesn't prove anything other than the MiSTer framework does indeed support a mouse!

The Arkanoid core is a bit more complicated, because it supports so many different controls - but by the looks of it the mouse part is very similar to the input test, just an integrator.

Centipede has the same-ish code from Arkanoid to integrate the position (only difference really being the clock speed it operates at), and then on top of that another layer to convert that to the format the Centipede hardware expects.  Annoying none of the Centipede schematics I can find actually show the mechanism that generates those signals!

The other issue we have here is that for some people with different hardware it is apparently working fine, so there can't be anything fundamentally wrong - hence why it feels like a timing thing, or some kind of speed-related difference.

Have you tried this device with Centipede on MAME at all?

**Comment by stiegis** [comment]
Hello!

I used to have mame and as far as I can remember Centipede worked quite well with an Ultimarc trackball.

Interesting sidenote here but far away from the topic .. I used to have a real Centipede arcade pcb and it was connected to an Ultimarc trackball - it showed kind of the same strange behaviour (would not move accurate and precisely week responsiveness overall weak controlability). So there seems to be some some curse lying upon us regarding this game (just kidding)

I still give not up because it is such a silly problem and I like this game so much. I am not buying a whole centipede cabinet just to get victory .. this is weak

However I am now onto the very weird and super comlicated configuration system of the Mister FPGA DE-10 nano. There is a button which is called Mouse Emu - if I get this one to be configurated I can try pressing it when centipede core has started. Perhaps this will toggle the right parameters and the trackball behaves like a real mouse finally.

By the way lets assume I want to test it with a wired computer mouse do I need one for windows 10 or does it have to be one especially for linux. I am also thinking about the linux mister fpga usb mouse.c driver to be a culprit for all this mess. Unfortunately I am no programmer but will not give up on this no way

regards,
Alex

**Comment by JimmyStones** [comment]
That's interesting about the original pcb having similar behaviour - in fact it's really useful info because it rules out a fair few possibilities.    I don't think the mouse emu function is relevant (although I've honestly never used it) as it is definitely receiving it as a mouse already (as seen in the input test).

So far, given all the other evidence I think it might be down to a difference in the the sensitivity level of the encoder between different devices - i.e. it's spitting out too much movement and that is overflowing the counters inside the 'wrapper' that links the Centipede core to the MiSTer framework.  I'll knock up a test build with some different values and send a link later on today.

Any USB mouse from the last 20 years will probably work!

**Comment by stiegis** [comment]
Hi!

hmmm interesting point with "overflow" values this could explain that the main character sometimes moves in the wrong direction - this happens when the trackball is moved some times and the character stops (like there is a resistance) you can go on rolling it will still stay and some time later it would go the opposite direction - there must be some form of "wrong" calculation going on.

This configuration thing in Mister is not a real help either .. 

Quote from mister joystick mapping "If you are defining keyboard keys for joystick emulation, then simply press the RIGHT arrow key and it will skip Stick ½ steps and jump to D-Pad keys definition" I mean this instructions does not make any sense at all and is far too complicated for a normal user

You would need a d-pad in the first place to define all the things but I have got only a joystick i-pack combination .. what if someone has connected an optipac with a trackball there is zero info about this

Also when you come to move mouse left point in the configuration you can roll the ball but it does not recognize only keyboard inputs .. 

You would need a normal trackingball configuration tool and when you move it it is correctly recognize as it was under windows in the old days but they talk about joystick emulation keyboard emulation this is so confusing to me

Don´t get me wrong the whole mister project is absolutely amazing but the implementation of controls is an absolute nightmare and desaster

thats all for now 

regards

**Comment by JimmyStones** [comment]
https://github.com/JimmyStones/Arcade-Centipede_MiSTer/blob/cocktail-fix/releases/Arcade-Centipede_20220410.rbf

Please give this one a try and see if there is any change!

MiSTer has a tough job catering for all users unfortunately, there are so many possible input devices and combinations of setup.  As it goes though, USB-based devices that present as a mouse should all just work out of the box with no configuration outside of that VIDPID stuff in the ini for certain devices.

**Comment by stiegis** [comment]
Hi!

Thx I will test this soon.

I also made a little step forward. I have just tested my setup with my computer mouse which uses a remote dongle. I thought it would not work at all but to my surprise it does!!!!!

Here comes the interesting part - the movement also feels kind of jerky and "out of control" the character sprite jumps from time to time and there is no smooth control as you would have on a leptop or desktop pc. I think this magnifies the problem - perhaps it is specific to my mister fpga. I am convinced that no matter which hardware (trackball type etc.) I connect to my setup this problem will persist so its software-hardware combination related).

I will try your new core next and also think about exchanging the otg usb hub where all the controls are connectet and also try to stick the opti pac usb connection in a different slot of the otg usb hub. Perhaps something funky is heaven withing this hardware.

By the way do you think the mra file has any effect upon the behaviour of the trackball movement?

regards,

**Comment by stiegis** [comment]
Hello Jimmy!

Just tried the new core and there is still no change. Same unpredictable unprecise behaviour of the trackball. For example the character sprite is standing at a point and I move in a certain direction to escape the spider it will make some random movement and crash into it. This is very frustrating and makes the game unplayable in the end. In this case it is better to play with a joystick.

Something very fundamental has to be wrong here perhaps the cpu itself is not implemented correctly or the centipede pcb has something special which is not covered at all. I can also think that centipede trackball one ore two are merged together or cocktail mode is somehow doing funky stuff in the core.  The reason this is so difficult to analyse is that it is hard to describe what is really happening on the screen - all you can say is it does not feel right at all - the control "derails" from time to time due to timing issues combined with some accidental behaviour (moves in the opposite direction of the movement of the trackball which normally would never ever happen)

thats all for now

regards
Alex

**Comment by JimmyStones** [comment]
I've been talking to other MiSTer users with trackballs and I'm acquiring one myself now that another user has and does have the same issue (always meant to get one anyway) so I will be able to work more easily on this once I have one to test directly with.  

Don't lose hope! 😁

**Comment by stiegis** [comment]
Hi Jimmy!

Yes this is a good idea - after all it is not like the ball is not working at all but it is not good  enough for a decent game of centipede. The fact that my modern mouse is working but the centipede sprite is kind of jumping around also brought me to the idea that this is also about resolution of the trackball (i.e sensitivity). If this is totally out of range weird things can happen but this is another educated guess. 

There are so  many things here in the chain the trackball with its own encoder wheel and pcb the ultimarc optipac the otg usb hub and last but not least the mister fpga which is a universe itself - therefore so many many many things can go wrong. Sometimes it is a very primitive cause but it is not simple since you also have to descover it.

Centipede is an absolute classic - if you can manage this it would mean the world to me but hey no problem here just take your time if it happens it happens ..

friendly greetings from Alex

**Comment by stiegis** [comment]
Just another thought of mine. It would be absolutely interesting to have a gameplay on the original arcade Centipede arcade cabinet and see how the trackball performs there. Would be intelligent to have a reference point here but I cannot imagine that what we are seing now is the original performance of the trackball motion
Its a pity we don´t have missile command for fpga now which would be a perfect canditate for comparison purposes .. thats all for now

regards,

**Comment by JimmyStones** [comment]
I played on an original Centipede cabinet 2 weeks ago!  It was definitely smoother and more precise than you have described -  when I get myself a trackball I will be able to properly compare.  I can always go back to the arcade for more 'research' :D

I am working on Missile Command as a core as well, but it's a tough one so it's not playable yet.

**Comment by stiegis** [comment]
Hi!

Yes I would definitely think so - in this state it is not really playable because the gameplay highly depends on the precise action of the trackball. 

Imagine a bent gun barrel - no way you would win a gun fight with

In the case of Centipede it is the poor ability of precise placing (stopping at the right x,y point) the centipede character and above of that the random movements/slowdowns  during gameplay which is absolutely deadly in the game

Perhaps we are talking about a simple driver issue here .. in windows often decives do not work properly unless faulty driver is removed but in this case we are facing FPGA and where exactly is this driver here. I exclude mechanical issues from the trackball (such as the number of slots in the encoder wheel which kind of represents sensitivity and number of signals being put out etc.) it MUST be a software thing in the end but of course there are also combined dynamic effects. The Suzzo Happ trackball is abolutely new and it works absolutely normal when connected to my laptop under Windows 10 x64 that is a fact

ggrrrrrrr if only we knew in which direction we should go for the problem

nice greetings

**Comment by stiegis** [comment]
just another thought of mine. Perhaps this issue is all about latency. For  "normal" arcade games you have the directions and push buttons. All buttons are based upon simple triggering a digital swich which of cours implies physical and software related latency. But in the end it is a simple digital switch on and off which is terminated. But in Centipede you have got kind of infinite row of analogue switches combinde which results in a definite physical motion at the end. In this chain all of the latency of each process is summed up which of course results into quite a bit of time which in the end messes up the counter in the fpga core. Just a theoretical assumption of mine

regards

**Comment by JimmyStones** [comment]
Good news - I now have a trackball, and it is just as bad as you describe in Centipede!  It also works fine as mouse in the Input Test core, and Arkanoid also - so I'm now in the same situation and can try some things to iterate on this problem now.  Will let you know how it goes 😁

**Comment by stiegis** [comment]
that is "good" news that the problem is the same with you therefore I know that my setup cannot be completely wrong

the fact that it is working quite good with Arkanoid would tell me that it  can only be a core issue there is only one little thing that has to be taken into consideration. 
In Arkanoid you only have a direction along the x-axis whereas in centipede it is also a direction along the y-axis.  Perhaps this is not implemented correctly in the core the values add up in the wrong way and cause the wrong positioning of the character sprite. At the end of the day its two spinners combined.

This is so weird and frustrating at the same time. Hope you get lucky with this one!!!

regards

**Comment by stiegis** [comment]
Hello Jimmy!

Any success with Centipede?

regards,

**Comment by JimmyStones** [comment]
Hallo - I've not had a lot of time to work on it so far (easter weekend = lots of family stuff) but I've ruled out some things at least...  I have more theories to test out but it's a bit slow progress because I have to compile the core every time and that takes about 6 mins even if you only change one line!  I'm sure it is solvable though :)

**Comment by stiegis** [comment]
Hello Jimmy!

Hope you can work it out but I guess this is not sure.

One thing is sure it must be the core itself or some "special" thing about the architecture of the centipede hardware that interferes with the rest of the environment of the mister. I have ordered the mister usb hub which is a nice addon and has an individual power suppy perhaps this is a current issue (little hope here but sometimes it is the things you would not expect at all)

The strangest thing about this is the fact that it was not working great at all with an original centipede arcade pcb .. I hooked up a brand new Ultimarc trackball directly to the joystick inputs of a mac supergun. It worked but really very poorly (I remember there was no wrong direction thing going on but also rolling the ball and movement only some time later thus poor responsiveness).

Considering the fact above the mister fpga core copies the problem in a way. So in conclusion it is possible that the core is ok but the fault lies between the enoder of the trackball itself and the encoder of the Ultimarc optipac. These two things are out of timing (not synchronous) or messing up each other)

These are the only realistic possibilities I can think of

However if this cannot be solve it is not the end of the world but a real shame

regards,

**Comment by stiegis** [comment]
Hello!

Found out that you can enter test mode in the atari centipede where you can move the trackball and press the fire button. The "pointer" also shows very weird behaviour, directions reversed, movement in the wrong direction. If you press the fire button a few times it kind of gets more precise. I don´t know if this is a kind of calibration thing or just for testing the controls.

In my opinion there are two main problems of the weird funky and non precise behaviour of the trackball. First of all it is lag (messing up the timings) and secondly it is very difficult (or even impossible) for mister fpga to mimic analog controls. In the case of the core input tester it is a digital linear concept therefore the mouse pointer is moving precise but always with a constant speed. In the real world you have different speeds accelerations which is solved by descrete electronic components on the actual arcade pcb. In my opinion Mister can only mimic the digital side of it but not the analog resulting in messed up playability of the trackball. With Joystick inputs it is a different story. You got digital switch on and switch off. Just a thought.

**Comment by JimmyStones** [comment]
It's all a bit odd - in the end of the day analog controls are all digital anyway to some degree!  In the case of the trackball, the movement is encoded as a series of pulses, and the faster the movement speed the more frequent the pulses.

Anyhow, some good news is that another dev who was previously working on this core has popped back up with some updates, so those are being tested now and might help.  If not, there is **another** dev who originally put the trackball support in who is up for taking a look next month after his current real life work calms down.  So there is hope!

**Comment by stiegis** [comment]
Hello!

I have not given hope up gg but this is really annoying as Centipede is an absolute classic.

We can speculate until the end of the world but this will not help. I guess it is such a primitive thing (if you know it) but it is hidden behind this intense complexitiy of details interacting with one another.

Whats funny is that joystick control works precisely so the core cannot be that bad it must be the interaction between the trackball encoder and the cpu (among other parts) which messes things up. I wonder where this "weird" stuff comes from - sometimes it does not react to the ball at all or suddenly turns the direction. This cannot be only a sensitivity thing - in the real pcb you would change sensitivity AND it would go in the right direction ALL the time (slower or faster) but in a predictable way. On the flip side sensitivity could ba a part of the problem causing this weird behaviour.

Perphaps we can see it happen in the near future

nice greetings

**Comment by stiegis** [comment]
Hello Jimmy!

Have you achieved any progress regarding trackball control with centipede? The "new" core is hidden under alternatives centipede. I tested it out and as far as vertical flip is concerned it works but mouse sensitivity feature is not really a huge step forward. It is far from being precise and smooth. I assume this will never happen so the only thing that is left is to buy an original centipede cab to get the real feeling. What a shame

regards,

**Comment by JimmyStones** [comment]
I haven't had any time to work on it any further I'm afraid - real work getting in the way unfortunately!  However, I have spoken to the developer who originally added the trackball support and he is going to help take a look at it in a couple of weeks after his own work project is over.
I don't think it is impossible to make this work, we'll definitely get there in the end 😅

**Comment by stiegis** [comment]
Hi!

Thanks for giving this a shot. It is certainly doable but I can only imagine how difficult this is. The whole mister fpga forum is full of issues and this is only one single problem showing how complex this old arcade pcbs are. To understand this from the electronic schematics is one thing but to implement this into a core is another story. If it happens one day fine otherwise there are more important things in life. In the end it is important what makes you happy.

regards,

**Comment by JaguarLV** [comment]
Try this to see if it is any better.

https://github.com/JaguarLV/Arcade-Centipede_MiSTer/blob/updated-mouse/trackball-sensitivity/releases/Arcade-Centipede_20220426.rbf

**Comment by stiegis** [comment]
Hello Jaguar!

Thx it is a little better than before but not even close to the real Centipede 

following issues in no particular order

shooter sometimes changes direction (unpredictable)
shooter sometimes gets stuck
shooter placement not precise 
shooter weird acceleration and therefore not precise
shooter kind of has micro wiggles and "jumps" from time to time not being smooth as for example computer mouse (resolution?)
movement feels sometimes better than again worse

impossible to play with such a fast paced game demanding high precision

trackball functions perfectly in core input tester

I wonder if Centiped has a hidden feature for the trackball (in the real game you can go into test) confusing mister´s fpga logic. 

When I connected it to a real centipede pcb I could not get it to work very good either (but it had not this weird things like changing directions) and it was not very responsive. One has to know how exactly Atari implemented this trackball. 

This trackball thing is really a sepecial one that is for sure.

Really funny in a way it is possible to write such a complex core but this type of control seems to have hidden secrets

thats life it is what it is 

regards

**Comment by James4739** [comment]
Just wanted to add to this interesting discussion to say I’m seeing the same behaviour. I have a MiSTer FPGA and X-Arcade trackball attached via USB. The control in Centipede is erratic, with the movement of the trackball not matching the on-screen movement. At times the character moves in the opposite direction to the trackball movement. The trackball works fine when set up as a mouse in other games on MiSTer, including the Arkanoid arcade core and The Irritating Maze on the Neo Geo console core. Centipede is an absolute classic and personal favourite, it would be so great to be able to play it as originally intended on the MiSTer using a trackball.

**Comment by stiegis** [comment]
Hello James!

It is nice that you noticed the issue as well. It would be absolutely amazing if the controls worked properly in this game. I do not quite understand the reason for the erratic behaviour because the spinner in Arkanoid works great and at the end of the day a trackball is a combination of two spinners with two axis x and y direction. The feature mouse sensitivity was added but this does not do any change at all. The core itself is really fine but the steering with the trackball is a nightmare. Not playable at all.

I think that mister is really great but its weak point are analogue controls. I bet we will never see a single game where the trackball works properly.  For example missile command would be another canditate where mister fails. I do not get it how can such complex cores be written on mister and it is not possible to get the controls right. Playing Centipede without a trackball is like driving a car with Joystick or buttons. That is not how it works and how it was meant to be played. Regards

**Comment by stiegis** [comment]
hi Jimmy!

Hurrrayyy Centipede is finally working - you are absolutely amazing!!! This is so amazing and I am very very very much impressed. You also did missile command. You are absolutely amazing. I would never thought this would happen COOL

**Comment by JimmyStones** [comment]
Haha, thanks! It took 3 months and another core but I got there in the end :D

Sounds like we can close this ticket 👍

**Comment by stiegis** [comment]
Hi Jimmy!

Thx for fixing Centipede man.  As a little sign of recognition I just bought a coffee for you. I am not quite sure if people are aware the mount of work which goes into this kind of stuff and how unbelievable great this is. Now my Centipede cab is perfect and you made it possible. I will support you work that is for sure.

nice greetings,
Alex
